basics/section_2-24.rb

```diff
@@ -0,0 +1,37 @@
+def operator(first_num, second_num, operation)
+  if operation == "multiply"
+    first_num * second_num
+  elsif operation == "divide"
+    first_num / second_num
+  elsif operation == "add"
+    first_num + second_num
+  elsif operation == "substract"
+    first_num - second_num
+  elsif operation == "modulus"
+    first_num % second_num
+  end
+end
+
+puts "Simple Calculator"
+20.times { print "-" }
+puts
+puts "Please enter the first number"
+num_1 = gets.chomp.to_f
+puts "Please enter the second number"
+num_2 = gets.chomp.to_f
+puts "Please enter the operation of choice"
+puts "multiply / divide / add / substract / modulus"
+operation = gets.chomp.downcase
+
+case operation
+when "multiply", "add", "substract"
+  puts "The first number #{operation} by the second number is #{operator(num_1, num_2, operation)}"
+when "divide", "modulus"
+  if num_2 == 0
+    puts "Cannot #{operation} #{num_1} by #{num_2}"
+  else
+    puts "The first number #{operation} by the second number is #{operator(num_1, num_2, operation)}"
+  end
+else
+  puts "Invalid operation"
+end
```

Comment:
```suggestion
def operator(first_num, second_num, operation)
  first_num = first_num.to_f
  second_num = second_num.to_f

  return first_num * second_num if operation == "multiply"
  return first_num + second_num if operation == "add"
  return first_num - second_num if operation == "substract"

  raise ZeroDivisionError.new "Cannot #{operation} #{first_num} by #{second_num}" if second_num == 0

  return first_num / second_num if operation == "divide"
  return first_num - second_num if operation == "modulus"

  puts "Invalid operation"
end

puts "Simple Calculator"
20.times { print "-" }
puts
puts "Please enter the first number"
num_1 = gets.chomp
puts "Please enter the second number"
num_2 = gets.chomp
puts "Please enter the operation of choice"
puts "multiply / divide / add / substract / modulus"
operation = gets.chomp.downcase

begin
  puts "The first number #{operation} by the second number is #{operator(num_1, num_2, operation)}"
rescue ZeroDivisionError => ex
  puts ex.message
end
```

This is how I'd do it.

1. A method should be a pure method (as much as possible)
2. return early to avoid if/else hell
3. Defensive programming: Assume the method consumer pass in wrong data format
4. To reduce a lot of unnecesary if-else, make use of design patterns

---

Either way, you choose. Once you ok, resolve this & merge.

Reply:
Thanks for the helpful example and pointers to keep in mind when trying to refactor the code. Is `ex` on line 30 short for `exception` and is that what's commonly used for naming the argument?

p.s. I've modified your code example a little to ignore line 29 when an invalid operation is called.

Reply:
I actually prefer my implementation because it keeps that within the method itself. This means if you call that method somewhere else, you would need to have that if else replicated again.

*I understand why you put it out. Cause if invalid, it will puts and you will have 2 puts. In that case, I will raise an exception there.

I did it in the manner above cause its a simple one. But, lets say the function is more complicated, this is how I would have done

```
puts "Simple Calculator"
20.times { print "-" }
puts
puts "Please enter the first number"
num_1 = gets.chomp
puts "Please enter the second number"
num_2 = gets.chomp
puts "Please enter the operation of choice"
puts "multiply / divide / add / substract / modulus"
operation = gets.chomp.downcase

begin
  puts "The first number #{operation} by the second number is #{calculate(num_1, num_2, operation)}"
rescue => ex
  puts ex.message
end
```

The above seems stupid to use `send`. This falls in the realm of meta programming. The reason we do this is because once we start working on classes & others that have different flow depending on what value is passed in, we can just do that without worrying about many if else. IF there is a new option, you can just add a new method and it'll work. No need an additional code on if -else